### PR TITLE
Improve `sprintf` return type inference

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/11201
closes https://github.com/phpstan/phpstan/issues/10493

### CODE REVIEW

File: tests/PHPStan/Analyser/nsrt/bug-7387.php [hunk 1]
```diff
@@ -8,60 +8,61 @@ class HelloWorld
...
-		assertType('numeric-string', sprintf('%2$14F', $mixed, $i));
-		assertType('numeric-string', sprintf('%2$14F', $mixed, $f));
-		assertType('numeric-string', sprintf('%2$14F', $mixed, $s));
+		assertType('non-falsy-string&numeric-string', sprintf('%2$14F', $mixed, $i));
+		assertType('non-falsy-string&numeric-string', sprintf('%2$14F', $mixed, $f));
+		assertType('non-falsy-string&numeric-string', sprintf('%2$14F', $mixed, $s));
 
-		assertType('string', sprintf('%10$14F', $mixed, $s));
+		// XXX should be string because of invalid arguments count
+		assertType('non-falsy-string&numeric-string', sprintf('%10$14F', $mixed, $s));
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/bug-7387.php:71]: after https://github.com/phpstan/phpstan-src/pull/3126 is merged we have the necessary `PrintfHelper` to decide about the invalid number of arguments/placeholders

File: src/Type/Php/SprintfFunctionDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -48,70 +47,85 @@ public function getTypeFromFunctionCall(
...
+			if ($functionReflection->getName() === 'vsprintf') {
+				if ($argType->isIterableAtLeastOnce()->yes()) {
+					$isNonEmpty = true;
+				}
+				continue;
+			}
 
-			if ($singlePlaceholderEarlyReturn !== null) {
-				return $singlePlaceholderEarlyReturn;
+			if ($argType->toString()->isNonFalsyString()->yes()) {
```
Comment (staabm) [inline src/Type/Php/SprintfFunctionDynamicReturnTypeExtension.php:116]: requires https://github.com/phpstan/phpstan-src/pull/3170

**Comment by staabm** [comment]
the composer error is valid, as we previously inferred a non-empty-string return type, even though the sprintf pattern is not known at static analysis time.

other errors are the same ones are unrelated

**Comment by staabm** [comment]
the reported phpstan error is interesstig.. it seems - depending on some unknown factor - reports that we are catching a non-throwable but ignoring this very same error by identifer leads to a "No error with identifier catch.notThrowable is reported on line 200."

see the 2 variants at
- https://github.com/phpstan/phpstan-src/pull/3168/commits/ff9189e5e14e3cbe3dc1b87eba1cd09763cb52da
- https://github.com/phpstan/phpstan-src/pull/3168/commits/fd5e76270f38ebf0e04137fff4426d4db264d7d7

(its unrelated to this PR though)

**Comment by staabm** [comment]
I think its related to this `Consistency::flexEntity` call in the exception classes.. some class-alias stuff is going on there.. this might lead to a race condition depending on the order which classes will be scanned when

**Comment by ondrejmirtes** [comment]
There's an opportunity to grab the code from Hoa packages and make a small single modern library with no magic and simple PSR-4 autoloading and replace Hoa in phpstan-src. I might look into it this summer.

**Comment by staabm** [comment]
> I might look into it this summer.

ok, what can we do to fix the build until then?

**Comment by ondrejmirtes** [comment]
Try to rebase on top of: https://github.com/phpstan/phpstan-src/commit/323315554048a22e6c01d635d6c5bb68f10d5aa1

**Comment by staabm** [comment]
> Try to rebase on top of: [3233155](https://github.com/phpstan/phpstan-src/commit/323315554048a22e6c01d635d6c5bb68f10d5aa1)

looks great, thanks

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
I'm sorry. I see the diff of `src/Type/Php/SprintfFunctionDynamicReturnTypeExtension.php` and my brain just freezes. I'm not able to review this. Can you do this in smaller steps, or maybe not move that many things around?

"Make the change easy, then make the easy change."

**Pull request closed without merging**